core/src/main/scala/org/bykn/bosatsu/Evaluation.scala

```diff
@@ -7,13 +7,104 @@ import cats.implicits._
 
 import org.bykn.bosatsu.rankn.Type
 
+object Evaluation {
+  import Value._
+
+
+  /**
+   * If we later determine that this performance matters
+   * and this wrapping is hurting, we could replace
+   * Value with a less structured type and put
+   * all the reflection into unapply calls but keep
+   * most of the API
+   */
+  sealed abstract class Value {
+    def asFn: FnValue =
+      this match {
+        case f@FnValue(_) => f
+        case other => sys.error(s"invalid cast to Fn: $other")
+      }
+  }
+
+  object Value {
+    sealed abstract class ProductValue extends Value {
+      def toList: List[Value] =
+        this match {
+          case UnitValue => Nil
+          case ConsValue(head, tail) => head :: tail.toList
+        }
+    }
+
+    object ProductValue {
+      def fromList(ps: List[Value]): ProductValue =
+        ps match {
+          case Nil => UnitValue
+          case h :: tail => ConsValue(h, fromList(tail))
+        }
+    }
+
+    case object UnitValue extends ProductValue
+    case class ConsValue(head: Value, tail: ProductValue) extends ProductValue
+    case class SumValue(variant: Int, value: ProductValue) extends Value
+    case class FnValue(toFn: Value => Eval[Value]) extends Value {
+      def apply(a: Value): Eval[Value] = toFn(a)
+    }
+    case class ExternalValue(toAny: Any) extends Value
+
+    val False: Value = SumValue(0, UnitValue)
+    val True: Value = SumValue(1, UnitValue)
+
+    object VInt {
+      def apply(v: Int): Value = ExternalValue(java.lang.Integer.valueOf(v))
+      def unapply(v: Value): Option[Int] =
+        v match {
+          case ExternalValue(v: java.lang.Integer) => Some(v.intValue)
+          case _ => None
+        }
+    }
+
+    object Str {
+      def apply(str: String): Value = ExternalValue(str)
+      def unapply(v: Value): Option[String] =
+        v match {
+          case ExternalValue(str: String) => Some(str)
+          case _ => None
+        }
+    }
+
+    object VOption {
+      def unapply(v: Value): Option[Option[Value]] =
+        v match {
+          case UnitValue =>
+            Some(None)
+          case ConsValue(head, UnitValue) =>
+            Some(Some(head))
+          case _ => None
+        }
+    }
+
+    object VList {
+      def unapply(v: Value): Option[List[Value]] =
```

Comment:
this should be a cons match: `Option[(Value, Value)]` to return head and tail, that is how we actually use it, in most cases.